File: trusera-sdk-js/src/cedar.ts
/**
 * Cedar-like policy evaluator for TypeScript.
 *
 * Evaluates simplified Cedar policy rules against HTTP request contexts.
 * Supports forbid/permit rules with conditions on URL, method, hostname, path.
 *
 * @example
 * ```typescript
 * const policy = `
 *   forbid (principal, action == Action::"http.get", resource)
 *   when { resource.hostname == "malicious.com" };
 * `;
 *
 * const evaluator = new CedarEvaluator(policy);
 * const result = evaluator.evaluate({
 *   url: "https://malicious.com/api",
 *   method: "GET",
 *   hostname: "malicious.com"
 * });
 *
 * console.log(result.decision); // "Deny"
 * console.log(result.reasons);  // ["Policy violation: hostname == malicious.com"]
 * ```
 */

/**
 * Request context for policy evaluation.
 */
export interface PolicyContext {
  /** Full URL being accessed */
  url: string;
  /** HTTP method (GET, POST, etc.) */
  method: string;
  /** Hostname extracted from URL */
  hostname: string;
  /** Path extracted from URL */
  path?: string;
  /** Custom context fields */
  [key: string]: unknown;
}

/**
 * Policy evaluation result.
 */
export interface PolicyDecision {
  /** Allow or Deny */
  decision: "Allow" | "Deny";
  /** Reasons for the decision (policy rules that matched) */
  reasons: string[];
}

/**
 * Parsed policy rule.
 */
interface PolicyRule {
  /** forbid or permit */
  effect: "forbid" | "permit";
  /** Action pattern (e.g., "http.get", "*") */
  action: string;
  /** Field name being checked (e.g., "hostname", "path") */
  field: string;
  /** Comparison operator (==, !=, contains, startsWith, etc.) */
  operator: string;
  /** Value to compare against */
  value: string;
  /** Raw rule text for debugging */
  raw: string;
}

/**
 * Cedar-like policy evaluator.
 *
 * Parses and evaluates simplified Cedar policies for HTTP requests.
 * Supports basic forbid/permit rules with string matching conditions.
 */
export class CedarEvaluator {
  private rules: PolicyRule[] = [];

  /**
   * Creates a new Cedar evaluator with the given policy text.
   *
   * @param policyText - Cedar-like policy rules
   */
  constructor(policyText: string) {
    this.rules = this.parsePolicy(policyText);
  }

  /**
   * Evaluates a request context against loaded policies.
   *
   * @param context - Request context to evaluate
   * @returns Policy decision with reasons
   */
  evaluate(context: PolicyContext): PolicyDecision {
    const matchedRules: PolicyRule[] = [];

    for (const rule of this.rules) {
      if (this.evaluateRule(rule, context)) {
        matchedRules.push(rule);
      }
    }

    // If any forbid rules matched, deny the request
    const forbidMatches = matchedRules.filter((r) => r.effect === "forbid");
    if (forbidMatches.length > 0) {
      return {
        decision: "Deny",
        reasons: forbidMatches.map(
          (r) => `Policy violation: ${r.field} ${r.operator} ${r.value}`
        ),
      };
    }

    // Otherwise allow
    return {
      decision: "Allow",
      reasons: [],
    };
  }

  /**
   * Parses Cedar-like policy text into rule objects.
   *
   * Supported syntax:
   * ```
   * forbid (principal, action == Action::"http.get", resource)
   * when { resource.hostname == "malicious.com" };
   *
   * forbid (principal, action == Action::"*", resource)
   * when { resource.url contains "blocked-domain.com" };
   * ```
   */
  private parsePolicy(policyText: string): PolicyRule[] {
    const rules: PolicyRule[] = [];

    // Strip comments (// style)
    const cleaned = policyText.replace(/\/\/[^\n]*/g, "");

    // Regex to match forbid/permit rules
    // Matches: (forbid|permit) ( ... ) when { resource.field operator "value" };
    const rulePattern = /\b(forbid|permit)\s*\(\s*principal\s*,\s*action\s*==\s*Action::"([^"]+)"\s*,\s*resource\s*\)\s*when\s*\{([^}]+)\}\s*;/gim;

    let match;
    while ((match = rulePattern.exec(cleaned)) !== null) {
      const effect = match[1]?.toLowerCase() as "forbid" | "permit";
      const action = match[2];
      const conditionBody = match[3];

      if (!effect || !action || !conditionBody) {
        continue;
      }

      // Parse condition (resource.field operator value)
      const conditionPattern = /resource\.(\w+)\s*(==|!=|contains|startsWith|endsWith)\s*"([^"]+)"/gi;
      let condMatch;
      while ((condMatch = conditionPattern.exec(conditionBody)) !== null) {
        const field = condMatch[1];
        const operator = condMatch[2];
        const value = condMatch[3];

        if (!field || !operator || !value) {
          continue;
        }

        rules.push({
          effect,
          action,
          field,
          operator,
          value,
          raw: match[0] ?? "",
        });
      }
    }

    return rules;
  }

  /**
   * Evaluates a single rule against a context.
   */
  private evaluateRule(rule: PolicyRule, context: PolicyContext): boolean {
    // Check if action matches
    if (!this.matchesAction(rule.action, context.method)) {
      return false;
    }

    // Get the field value from context
    const fieldValue = this.getFieldValue(rule.field, context);
    if (fieldValue === undefined) {
      return false;
    }

    // Evaluate the condition
    return this.evaluateCondition(
      String(fieldValue),
      rule.operator,
      rule.value
    );
  }

  /**
   * Checks if an action pattern matches the HTTP method.
   */
  private matchesAction(actionPattern: string, method: string): boolean {
    if (actionPattern === "*") {
      return true;
    }

    // Action format: "http.get", "http.post", etc.
    const normalizedPattern = actionPattern.toLowerCase();
    const normalizedMethod = method.toLowerCase();

    if (normalizedPattern === `http.${normalizedMethod}`) {
      return true;
    }

    // Check for wildcard patterns
    if (normalizedPattern.endsWith("*")) {
      const prefix = normalizedPattern.slice(0, -1);
      return normalizedMethod.startsWith(prefix);
    }

    return false;
  }

  /**
   * Extracts a field value from the context.
   */
  private getFieldValue(field: string, context: PolicyContext): unknown {
    // Direct context access
    if (field in context) {
      return context[field];
    }

    // Special handling for common fields
    if (field === "hostname") {
      return context.hostname;
    }

    if (field === "path") {
      if (context.path !== undefined) {
        return context.path;
      }
      // Try to extract from URL
      try {
        const url = new URL(context.url);
        return url.pathname;
      } catch {
        return undefined;
      }
    }

    if (field === "method") {
      return context.method;
    }

    if (field === "url") {
      return context.url;
    }

    return undefined;
  }

  /**
   * Evaluates a condition operator.
   */
  private evaluateCondition(
    actual: string,
    operator: string,
    expected: string
  ): boolean {
    const actualLower = actual.toLowerCase();
    const expectedLower = expected.toLowerCase();

    switch (operator.toLowerCase()) {
      case "==":
        return actualLower === expectedLower;
      case "!=":
        return actualLower !== expectedLower;
      case "contains":
        return actualLower.includes(expectedLower);
      case "startswith":
        return actualLower.startsWith(expectedLower);
      case "endswith":
        return actualLower.endsWith(expectedLower);
      default:
        return false;
    }
  }

  /**
   * Returns the number of loaded rules.
   */
  getRuleCount(): number {
    return this.rules.length;
  }
}

File: trusera-agent-sdk/trusera_sdk/cedar.py
"""
Embedded Cedar-like policy evaluator for HTTP request filtering.

This module provides a lightweight Cedar policy engine adapted from the AI-BOM
Cedar gate for evaluating HTTP requests against security policies.

Supported policy patterns:
  - forbid ... when { request.url contains "api.deepseek.com" };
  - forbid ... when { request.hostname == "deepseek.com" };
  - forbid ... when { request.method == "POST" };
  - forbid ... when { request.path contains "/upload" };
  - permit ... when { request.hostname == "api.openai.com" };

Policy evaluation precedence:
  1. Explicit forbid rules are evaluated first
  2. Explicit permit rules are evaluated second
  3. If no rules match, default to ALLOW
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum
from urllib.parse import urlparse


class PolicyAction(str, Enum):
    """Policy rule actions."""

    FORBID = "forbid"
    PERMIT = "permit"


class PolicyDecision(str, Enum):
    """Policy evaluation decision."""

    ALLOW = "allow"
    DENY = "deny"


@dataclass
class PolicyRule:
    """A single parsed Cedar-like policy rule for HTTP requests."""

    action: PolicyAction
    field: str
    operator: str
    value: str
    raw: str


@dataclass
class EvaluationResult:
    """Result of evaluating a request against policies."""

    decision: PolicyDecision
    reason: str
    matched_rule: PolicyRule | None = None


# Regex patterns for Cedar-like policy syntax
# Matches: forbid ( principal, action == Action::"http", resource ) when { ... };
RULE_PATTERN = re.compile(
    r'(forbid|permit)\s*\(\s*principal\s*,\s*action\s*==\s*Action::"(\w+)"\s*,\s*resource\s*\)'
    r"\s*when\s*\{([^}]+)\}\s*;",
    re.MULTILINE | re.DOTALL | re.IGNORECASE,
)

# Matches conditions inside when { ... }
# e.g. request.url contains "api.deepseek.com"  or  request.method == "POST"
CONDITION_PATTERN = re.compile(
    r'request\.(\w+)\s*(==|!=|contains|startswith|endswith)\s*"([^"]+)"',
    re.MULTILINE | re.IGNORECASE,
)


def parse_policy(policy_text: str) -> list[PolicyRule]:
    """
    Parse a Cedar-like policy file into a list of rules.

    Args:
        policy_text: Cedar policy text content

    Returns:
        List of parsed PolicyRule objects
    """
    rules: list[PolicyRule] = []
    # Strip comments (// style)
    cleaned = re.sub(r"//[^\n]*", "", policy_text)

    for match in RULE_PATTERN.finditer(cleaned):
        action_str = match.group(1).lower()
        action = PolicyAction.FORBID if action_str == "forbid" else PolicyAction.PERMIT
        body = match.group(3).strip()

        for cond in CONDITION_PATTERN.finditer(body):
            field_name = cond.group(1).lower()
            operator = cond.group(2).lower()
            value = cond.group(3)

            rules.append(
                PolicyRule(
                    action=action,
                    field=field_name,
                    operator=operator,
                    value=value,
                    raw=match.group(0).strip(),
                )
            )

    return rules


def _match_condition(rule: PolicyRule, value: str) -> bool:
    """
    Check if a value matches a rule condition.

    Args:
        rule: The policy rule to evaluate
        value: The actual value to compare

    Returns:
        True if the condition matches (violates for forbid, allows for permit)
    """
    rule_value = rule.value.lower()
    value_lower = value.lower()

    if rule.operator == "==":
        return value_lower == rule_value
    elif rule.operator == "!=":
        return value_lower != rule_value
    elif rule.operator == "contains":
        return rule_value in value_lower
    elif rule.operator == "startswith":
        return value_lower.startswith(rule_value)
    elif rule.operator == "endswith":
        return value_lower.endswith(rule_value)

    return False


def evaluate_request(
    url: str,
    method: str,
    headers: dict[str, str] | None = None,
    rules: list[PolicyRule] | None = None,
) -> EvaluationResult:
    """
    Evaluate an HTTP request against Cedar policy rules.

    Evaluation logic:
    1. Check all FORBID rules - if any match, DENY
    2. Check all PERMIT rules - if any match, ALLOW
    3. If no rules match, default to ALLOW

    Args:
        url: The request URL
        method: HTTP method (GET, POST, etc.)
        headers: Optional request headers
        rules: List of policy rules to evaluate (empty = allow all)

    Returns:
        EvaluationResult with decision and reason
    """
    if not rules:
        return EvaluationResult(
            decision=PolicyDecision.ALLOW,
            reason="No policy rules configured",
        )

    parsed = urlparse(url)
    hostname = parsed.hostname or ""
    path = parsed.path or "/"

    # Build request context for evaluation
    request_data = {
        "url": url,
        "method": method.upper(),
        "hostname": hostname,
        "path": path,
        "scheme": parsed.scheme,
    }

    # Add headers to context if provided
    if headers:
        for key, value in headers.items():
            request_data[f"header_{key.lower().replace('-', '_')}"] = value

    # Phase 1: Check FORBID rules (highest priority)
    for rule in rules:
        if rule.action != PolicyAction.FORBID:
            continue

        field_value = request_data.get(rule.field, "")
        if field_value and _match_condition(rule, field_value):
            return EvaluationResult(
                decision=PolicyDecision.DENY,
                reason=(
                    f'Forbidden by policy: request.{rule.field} {rule.operator} "{rule.value}"'
                ),
                matched_rule=rule,
            )

    # Phase 2: Check PERMIT rules (second priority)
    for rule in rules:
        if rule.action != PolicyAction.PERMIT:
            continue

        field_value = request_data.get(rule.field, "")
        if field_value and _match_condition(rule, field_value):
            return EvaluationResult(
                decision=PolicyDecision.ALLOW,
                reason=(
                    f"Explicitly permitted by policy: request.{rule.field} "
                    f'{rule.operator} "{rule.value}"'
                ),
                matched_rule=rule,
            )

    # Default: Allow if no rules matched
    return EvaluationResult(
        decision=PolicyDecision.ALLOW,
        reason="No matching policy rules (default allow)",
    )


class CedarEvaluator:
    """
    Cedar policy evaluator for HTTP requests.

    This class loads and evaluates Cedar-like policies against HTTP requests,
    providing a decision (allow/deny) and reasoning.

    Example:
        >>> evaluator = CedarEvaluator.from_file("policy.cedar")
        >>> result = evaluator.evaluate("https://api.deepseek.com/v1/chat", "POST")
        >>> if result.decision == PolicyDecision.DENY:
        ...     print(f"Blocked: {result.reason}")
    """

    def __init__(self, rules: list[PolicyRule]) -> None:
        """
        Initialize the evaluator with parsed rules.

        Args:
            rules: List of PolicyRule objects
        """
        self.rules = rules

    @classmethod
    def from_file(cls, policy_path: str) -> CedarEvaluator:
        """
        Load a Cedar policy from a file.

        Args:
            policy_path: Path to the .cedar policy file

        Returns:
            CedarEvaluator instance

        Raises:
            FileNotFoundError: If policy file doesn't exist
        """
        with open(policy_path, encoding="utf-8") as f:
            policy_text = f.read()

        rules = parse_policy(policy_text)
        return cls(rules)

    @classmethod
    def from_text(cls, policy_text: str) -> CedarEvaluator:
        """
        Load a Cedar policy from text.

        Args:
            policy_text: Cedar policy text

        Returns:
            CedarEvaluator instance
        """
        rules = parse_policy(policy_text)
        return cls(rules)

    def evaluate(
        self,
        url: str,
        method: str = "GET",
        headers: dict[str, str] | None = None,
    ) -> EvaluationResult:
        """
        Evaluate an HTTP request against loaded policies.

        Args:
            url: The request URL
            method: HTTP method (default: GET)
            headers: Optional request headers

        Returns:
            EvaluationResult with decision and reason
        """
        return evaluate_request(url, method, headers, self.rules)
